### PR TITLE
Master zdb allocation

### PR DESCRIPTION
**Pull request by muhamadazmy**
## Changes
- ZDB subvolume has no limit set on it
- Each namespace information is read from the zdb info file on disk 
- On subvolume creation, the system is scanned for the namespace on disk, if found, no suvol allocation is done
- Ensure that the container is running
- Ensure that the namespace is created in zdb

### CODE REVIEW

File: pkg/provision/zdb.go [hunk 1]
```diff
@@ -299,13 +207,12 @@ func createZdbContainer(ctx context.Context, name string, mode pkg.ZDBMode, volu
 	bo := backoff.NewExponentialBackOff()
 	bo.MaxInterval = time.Minute * 2
 	if err := backoff.RetryNotify(getIP, bo, nil); err != nil {
```
Comment (zaibon) [inline pkg/provision/zdb.go]:
```suggestion
	if err := backoff.Retry(getIP, bo); err != nil {
```

File: pkg/provision/zdb.go [hunk 2]
```diff
@@ -358,11 +268,14 @@ func zdbDecommission(ctx context.Context, reservation *Reservation) error {
 		return errors.Wrap(err, "failed to decode reservation schema")
 	}
 
-	containerID, ok := zdbMap.Get(nsID)
-	if !ok {
+	allocation, err := storage.Find(reservation.ID)
```
Comment (zaibon) [inline pkg/provision/zdb.go:301]: if Find return not found here, does that mean the 0-db namespace is also gone ? Cause do an early return then.

File: pkg/storage/zdb.go [hunk 3]
```diff
@@ -36,116 +70,120 @@ func (s *storageModule) Allocate(diskType pkg.DeviceType, size uint64, mode pkg.
...
-				v = volume
-				break
+			// skip all non-zdb volume
+			if !isZdbVolume(volume) {
+				continue
+			}
+
+			zdb := zdbpool.New(volume.Path())
+
+			reserved, err := zdb.Reserved()
```
Comment (zaibon) [inline pkg/storage/zdb.go]: Here again the way we compute the reserved storage of a disk is not correct. In this case we only check the amount of storage used by 0-db namespaces but we omit the storage reserved by container volumes.

**Review by zaibon — CHANGES_REQUESTED** [review]

File: pkg/provision/zdb.go (hunk 2 above)
Comment (muhamadazmy) [reply pkg/provision/zdb.go:301]: It means the namespace directory in 0-db data is gone. I am pretty sure it also means the namespace is gone from zdb itself.

Hence we do an early return once it's 'not found'.

File: pkg/storage/zdb.go (hunk 3 above)
Comment (muhamadazmy) [reply pkg/storage/zdb.go]: The 0-db subvolume is created with "unlimited" space (size == 0). Hence the only way to compute this is by using the zdbpool reserved.

Yesterday I was considering actually make the btrfs a bit smarter, and that they use the zdbpool to compute the "reserved" space of a subvolume if it is a zdb subvolume. In that case even subvolume creations will take namespaces reservations into account. What do u think? should we do that?

Comment (zaibon) [reply pkg/storage/zdb.go]: yes we should

File: pkg/storage/storage.go [hunk 4]
```diff
@@ -238,6 +239,9 @@ func (s *storageModule) Maintenance() error {
 // CreateFilesystem with the given size in a storage pool.
 func (s *storageModule) CreateFilesystem(name string, size uint64, poolType pkg.DeviceType) (string, error) {
 	log.Info().Msgf("Creating new volume with size %d", size)
+	if strings.HasPrefix(name, "zdb") {
```
Comment (zaibon) [inline pkg/storage/storage.go:242]: :+1:

**Review by zaibon — APPROVED** [review]
LGTM

File: pkg/provision/zdb.go [hunk 5]
```diff
@@ -193,6 +193,25 @@ func createZdbContainer(ctx context.Context, allocation pkg.Allocation, mode pkg
 		return errors.Wrap(err, "failed to create container")
 	}
 
+	socket := fmt.Sprintf("unix://%s/zdb.sock", socketDir)
```
Comment (zaibon) [inline pkg/provision/zdb.go]: let's extract this into a function now that you use it in different places.

Comment (zaibon) [inline pkg/provision/zdb.go]: let's extract this into a function now that you use it in different places.

File: pkg/provision/zdb.go [hunk 6]
```diff
@@ -193,6 +193,25 @@ func createZdbContainer(ctx context.Context, allocation pkg.Allocation, mode pkg
...
+	cl := zdb.New(socket)
+	defer cl.Close()
+
+	check := func() error {
```
Comment (zaibon) [inline pkg/provision/zdb.go]: you don't need to wrap it here. just use `cl.Connect` in the `RetryNotify`

**Review by zaibon — CHANGES_REQUESTED** [review]

File: pkg/provision/zdb.go (hunk 6 above)
Comment (muhamadazmy) [reply pkg/provision/zdb.go]: Yup, true. :stuck_out_tongue_winking_eye:

**Review by zaibon — APPROVED** [review]

**Pull request merged**